src/static/js/chapter.js

```diff
@@ -0,0 +1,189 @@
+    if (dataSaver) {
+      console.log('DataSaver is enabled');
+      gtag('event', 'data-saver', { 'event_category': 'user', 'event_label': 'enabled', 'value': 1 });
+    } else {
+      gtag('event', 'data-saver', { 'event_category': 'user', 'event_label': 'not-enabled', 'value': 0 });
+    }
+  }
+
+  return dataSaver;
+}
+
+//iOS causes Google Sheets to create a 6000 by 3700 canvas, which annoyingly isn't supported by iOS!
+//Let's check if we have large Canvas support (annoyingly no API for this!)
+function highResolutionCanvasSupported() {
+
+  var largeCanvasSupported = false;
+
+  try {
+    // Set large sized canvas dimensions and draw test rectangle
+    var cvs = document ? document.createElement('canvas') : null;
+    var ctx = cvs && cvs.getContext ? cvs.getContext('2d') : null;
+    cvs.width = 6000;
+    cvs.height = 3700;
+    ctx.fillRect(5999, 3699, 1, 1);
+    largeCanvasSupported = Boolean(ctx.getImageData(5999, 3699, 1, 1).data[3]) == 1;
+  }
+  catch (e) {
+    largeCanvasSupported = false;
+  }
+
+  if (largeCanvasSupported) {
+    gtag('event', 'hi-res-canvas', { 'event_category': 'user', 'event_label': 'supported', 'value': 1 });
+  } else {
+    console.log('High resolution canvas images are not supported');
+    gtag('event', 'hi-res-canvas', { 'event_category': 'user', 'event_label': 'not-supported', 'value': 0 });
+  }
+
+  return largeCanvasSupported;
+
+};
+
+//If google sheets test pixel works, then can assume interactive sheets work and can remove it
+function googleSheetsPixelLoaded() {
+  this.parentElement.removeChild(this);
+  gtag('event', 'sheets-access', { 'event_category': 'user', 'event_label': 'successful', 'value': 1 });
+  gtag('event', 'interactive-figures', { 'event_category': 'user', 'event_label': 'enabled', 'value': 1 });
+}
+
+//If google sheets test pixel doesn't work, then revert back to static images
+function googleSheetsPixelNotLoaded() {
+  console.error('Google Sheets cannot be loaded');
+
+  this.parentElement.removeChild(this);
+
+  var all_fig_imgs = document.querySelectorAll('figure .fig-mobile');
+  for (index = 0; index < all_fig_imgs.length; ++index) {
+    var fig_img = all_fig_imgs[index];
+    fig_img.classList.remove("fig-mobile");
+  }
+
+  var all_fig_iframes = document.querySelectorAll('figure .fig-iframe');
+  for (index = 0; index < all_fig_iframes.length; ++index) {
+    var fig_iframe = all_fig_iframes[index];
+    fig_iframe.parentElement.removeChild(fig_iframe);
+  }
+  gtag('event', 'google-sheets', { 'event_category': 'user', 'event_label': 'blocked', 'value': 1 });
+  gtag('event', 'interactive-figures', { 'event_category': 'user', 'event_label': 'not-enabled', 'value': 1 });
+
+}
+
+//We use Google Sheets for detailed visualisations
+//Check for support and switch out images if supported
+function upgradeInteractiveFigures() {
+
+  try {
+    if (bigEnoughForInteractiveFigures() && highResolutionCanvasSupported() && !dataSaverEnabled()) {

```

Comment:
Maybe do the `dataSaverEnabled()` check first, so that the condition can short-circuit quickly of it is enabled? If someone has dataSaver enabled, chances are that they may not be running on the best hardware, and the rest of the checks will likely return false. (This is pure conjecture.)

Reply:
Great idea - done!